Error compiling on arch/artix linux with latest version of libmicrohttpd

I'am trying to compile the AUR package of kiwix-lib but it throws an error while compiling:
```bash
templates/captured_external.html ['templates', 'captured_external_html']
opensearchdescription.xml ['opensearchdescription_xml']
[24/51] Compiling C++ object 'src/25a6634@@kiwix@sha/server.cpp.o'
FAILED: src/25a6634@@kiwix@sha/server.cpp.o 
c++ -Isrc/25a6634@@kiwix@sha -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/include/p11-kit-1 -flto -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -march=znver1 -m64 -mtune=generic -O2 -fno-plt -D_FORTIFY_SOURCE=2 -fPIC -pthread -MD -MQ 'src/25a6634@@kiwix@sha/server.cpp.o' -MF 'src/25a6634@@kiwix@sha/server.cpp.o.d' -o 'src/25a6634@@kiwix@sha/server.cpp.o' -c ../src/server.cpp
../src/server.cpp: In member function ‘bool kiwix::InternalServer::start()’:
../src/server.cpp:248:29: error: invalid conversion from ‘int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, size_t*, void**)’ {aka ‘int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, long unsigned int*, void**)’} to ‘MHD_AccessHandlerCallback’ {aka ‘MHD_Result (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, long unsigned int*, void**)’} [-fpermissive]
  248 |                             &staticHandlerCallback,
      |                             ^~~~~~~~~~~~~~~~~~~~~~
      |                             |
      |                             int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, size_t*, void**) {aka int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, long unsigned int*, void**)}
In file included from ../src/server.cpp:39:
/usr/include/microhttpd.h:2428:45: note:   initializing argument 5 of ‘MHD_Daemon* MHD_start_daemon(unsigned int, uint16_t, MHD_AcceptPolicyCallback, void*, MHD_AccessHandlerCallback, void*, ...)’
 2428 |                   MHD_AccessHandlerCallback dh, void *dh_cls,
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~^~
[35/51] Compiling C++ object 'src/25a6634@@kiwix@sha/server_request_context.cpp.o'
FAILED: src/25a6634@@kiwix@sha/server_request_context.cpp.o 
c++ -Isrc/25a6634@@kiwix@sha -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/include/p11-kit-1 -flto -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -march=znver1 -m64 -mtune=generic -O2 -fno-plt -D_FORTIFY_SOURCE=2 -fPIC -pthread -MD -MQ 'src/25a6634@@kiwix@sha/server_request_context.cpp.o' -MF 'src/25a6634@@kiwix@sha/server_request_context.cpp.o.d' -o 'src/25a6634@@kiwix@sha/server_request_context.cpp.o' -c ../src/server/request_context.cpp
../src/server/request_context.cpp: In constructor ‘kiwix::RequestContext::RequestContext(MHD_Connection*, std::string, const string&, const string&, const string&)’:
../src/server/request_context.cpp:79:58: error: invalid conversion from ‘int (*)(void*, MHD_ValueKind, const char*, const char*)’ to ‘MHD_KeyValueIterator’ {aka ‘MHD_Result (*)(void*, MHD_ValueKind, const char*, const char*)’} [-fpermissive]
   79 |   MHD_get_connection_values(connection, MHD_HEADER_KIND, &RequestContext::fill_header, this);
      |                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                          |
      |                                                          int (*)(void*, MHD_ValueKind, const char*, const char*)
In file included from ../src/server/request_context.h:33,
                 from ../src/server/request_context.cpp:22:
/usr/include/microhttpd.h:2699:49: note:   initializing argument 3 of ‘int MHD_get_connection_values(MHD_Connection*, MHD_ValueKind, MHD_KeyValueIterator, void*)’
 2699 |                            MHD_KeyValueIterator iterator,
      |                            ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
../src/server/request_context.cpp:80:64: error: invalid conversion from ‘int (*)(void*, MHD_ValueKind, const char*, const char*)’ to ‘MHD_KeyValueIterator’ {aka ‘MHD_Result (*)(void*, MHD_ValueKind, const char*, const char*)’} [-fpermissive]
   80 |   MHD_get_connection_values(connection, MHD_GET_ARGUMENT_KIND, &RequestContext::fill_argument, this);
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                |
      |                                                                int (*)(void*, MHD_ValueKind, const char*, const char*)
In file included from ../src/server/request_context.h:33,
                 from ../src/server/request_context.cpp:22:
/usr/include/microhttpd.h:2699:49: note:   initializing argument 3 of ‘int MHD_get_connection_values(MHD_Connection*, MHD_ValueKind, MHD_KeyValueIterator, void*)’
 2699 |                            MHD_KeyValueIterator iterator,
      |                            ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
[37/51] Compiling C++ object 'test/9f86d08@@server@exe/server.cpp.o'
ninja: build stopped: subcommand failed.
==> ERROR: A failure occurred in build().
    Aborting...
error making: %!s(func() string=0x55de324d5450)
```

The same error appears while manually compiling (the newest two releases).

I use the newest version of Artix Linux and would like to use kiwix

I tried to compile it manually using the python scripts from the repository you mentioned. If I run `kiwix-build kiwix-desktop` I get following message:

```bash
ERROR: The target platform native_dyn cannot be build on host deepin.
```

It continues though.
Building the kiwix-lib using this script worked but compiling the desktop version fails while compiling qt4:
```bash
      |                               ~~~~~~~~~~~~^~~~
/usr/include/c++/10.1.0/bits/random.h:518:7: note: candidate: ‘std::mersenne_twister_engine<_UIntType, __w, __n, __m, __r, __a, __u, __d, __s, __b, __t, __c, __l, __f>::mersenne_twister_engine() [with _UIntType = unsigned int; long unsigned int __w = 32; long unsigned int __n = 624; long unsigned int __m = 397; long unsigned int __r = 31; _UIntType __a = 2567483615; long unsigned int __u = 11; _UIntType __d = 4294967295; long unsigned int __s = 7; _UIntType __b = 2636928640; long unsigned int __t = 15; _UIntType __c = 4022730752; long unsigned int __l = 18; _UIntType __f = 1812433253]’
  518 |       mersenne_twister_engine() : mersenne_twister_engine(default_seed) { }
      |       ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/10.1.0/bits/random.h:518:7: note:   candidate expects 0 arguments, 1 provided
/usr/include/c++/10.1.0/bits/random.h:463:11: note: candidate: ‘constexpr std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>::mersenne_twister_engine(const std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>&)’
  463 |     class mersenne_twister_engine
      |           ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/10.1.0/bits/random.h:463:11: note:   no known conversion for argument 1 from ‘QRandomGenerator::SystemGenerator’ to ‘const std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>&’
/usr/include/c++/10.1.0/bits/random.h:463:11: note: candidate: ‘constexpr std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>::mersenne_twister_engine(std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>&&)’
/usr/include/c++/10.1.0/bits/random.h:463:11: note:   no known conversion for argument 1 from ‘QRandomGenerator::SystemGenerator’ to ‘std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>&&’
make: *** [Makefile:318: qrandom.o] Error 1
```




Compiling the desktop repository manually exits with following error:
```bash
/usr/bin/ld: warning: libicui18n.so.67, needed by /usr/lib/libQt5WebEngineCore.so, may conflict with libicui18n.so.58
/usr/bin/ld: warning: libicuuc.so.67, needed by /usr/lib/libQt5WebEngineCore.so, may conflict with libicuuc.so.58
/usr/bin/ld: warning: libicudata.so.67, needed by /usr/lib/libicuuc.so.67, may conflict with libicudata.so.58
/usr/bin/ld: /usr/lib/libpng16.so.16: undefined reference to `inflateValidate@ZLIB_1.2.9'
collect2: error: ld returned 1 exit status
make: *** [Makefile:552: kiwix-desktop] Error 1
```

Build error on Fedora:

```c++
[14/51] g++ -Isrc/libkiwix.so.9.3.0.p -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/include/p11-kit-1 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -O2 -fexceptions -g -grecord-gcc-switches -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -pthread -MD -MQ src/libkiwix.so.9.3.0.p/server.cpp.o -MF src/libkiwix.so.9.3.0.p/server.cpp.o.d -o src/libkiwix.so.9.3.0.p/server.cpp.o -c ../src/server.cpp
FAILED: src/libkiwix.so.9.3.0.p/server.cpp.o 
g++ -Isrc/libkiwix.so.9.3.0.p -Isrc -I../src -Iinclude -I../include -Istatic -I/usr/include/p11-kit-1 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -O2 -fexceptions -g -grecord-gcc-switches -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -pthread -MD -MQ src/libkiwix.so.9.3.0.p/server.cpp.o -MF src/libkiwix.so.9.3.0.p/server.cpp.o.d -o src/libkiwix.so.9.3.0.p/server.cpp.o -c ../src/server.cpp
../src/server.cpp: In member function ‘bool kiwix::InternalServer::start()’:
../src/server.cpp:248:29: error: invalid conversion from ‘int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, size_t*, void**)’ {aka ‘int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, long unsigned int*, void**)’} to ‘MHD_AccessHandlerCallback’ {aka ‘MHD_Result (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, long unsigned int*, void**)’} [-fpermissive]
  248 |                             &staticHandlerCallback,
      |                             ^~~~~~~~~~~~~~~~~~~~~~
      |                             |
      |                             int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, size_t*, void**) {aka int (*)(void*, MHD_Connection*, const char*, const char*, const char*, const char*, long unsigned int*, void**)}
In file included from ../src/server.cpp:39:
/usr/include/microhttpd.h:2428:45: note:   initializing argument 5 of ‘MHD_Daemon* MHD_start_daemon(unsigned int, uint16_t, MHD_AcceptPolicyCallback, void*, MHD_AccessHandlerCallback, void*, ...)’
 2428 |                   MHD_AccessHandlerCallback dh, void *dh_cls,
      |                   ~~~~~~~~~~~~~~~~~~~~~~~~~~^~
```

--- a/src/microhttpd_wrapper.h
+++ b/src/microhttpd_wrapper.h
@@ -1,0 +1,24 @@
+/*
+ * Copyright 2020 Emmanuel Engelhart <kelson@kiwix.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU  General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#include <microhttpd.h>
+
+#if MHD_VERSION < 0x00097002
+typedef int MHD_Result;
+#endif

--- a/src/server.cpp
+++ b/src/server.cpp
@@ -36,7 +36,7 @@
 #endif
 
 extern "C" {
-#include <microhttpd.h>
+#include "microhttpd_wrapper.h"
 }
 
 #include "tools/otherTools.h"
@@ -77,14 +77,14 @@ static IdNameMapper defaultNameMapper;
 
 typedef kainjow::mustache::data MustacheData;
 
-static int staticHandlerCallback(void* cls,
-                                 struct MHD_Connection* connection,
-                                 const char* url,
-                                 const char* method,
-                                 const char* version,
-                                 const char* upload_data,
-                                 size_t* upload_data_size,
-                                 void** cont_cls);
+static MHD_Result staticHandlerCallback(void* cls,
+                                        struct MHD_Connection* connection,
+                                        const char* url,
+                                        const char* method,
+                                        const char* version,
+                                        const char* upload_data,
+                                        size_t* upload_data_size,
+                                        void** cont_cls);
 
 
 class InternalServer {
@@ -101,13 +101,13 @@ class InternalServer {
                    bool blockExternalLinks);
     virtual ~InternalServer() = default;
 
-    int handlerCallback(struct MHD_Connection* connection,
-                        const char* url,
-                        const char* method,
-                        const char* version,
-                        const char* upload_data,
-                        size_t* upload_data_size,
-                        void** cont_cls);
+    MHD_Result handlerCallback(struct MHD_Connection* connection,
+                               const char* url,
+                               const char* method,
+                               const char* version,
+                               const char* upload_data,
+                               size_t* upload_data_size,
+                               void** cont_cls);
     bool start();
     void stop();
 
@@ -267,14 +267,14 @@ void InternalServer::stop()
   MHD_stop_daemon(mp_daemon);
 }
 
-static int staticHandlerCallback(void* cls,
-                                 struct MHD_Connection* connection,
-                                 const char* url,
-                                 const char* method,
-                                 const char* version,
-                                 const char* upload_data,
-                                 size_t* upload_data_size,
-                                 void** cont_cls)
+static MHD_Result staticHandlerCallback(void* cls,
+                                        struct MHD_Connection* connection,
+                                        const char* url,
+                                        const char* method,
+                                        const char* version,
+                                        const char* upload_data,
+                                        size_t* upload_data_size,
+                                        void** cont_cls)
 {
   InternalServer* _this = static_cast<InternalServer*>(cls);
 
@@ -287,13 +287,13 @@ static int staticHandlerCallback(void* cls,
                                 cont_cls);
 }
 
-int InternalServer::handlerCallback(struct MHD_Connection* connection,
-                                    const char* url,
-                                    const char* method,
-                                    const char* version,
-                                    const char* upload_data,
-                                    size_t* upload_data_size,
-                                    void** cont_cls)
+MHD_Result InternalServer::handlerCallback(struct MHD_Connection* connection,
+                                           const char* url,
+                                           const char* method,
+                                           const char* version,
+                                           const char* upload_data,
+                                           size_t* upload_data_size,
+                                           void** cont_cls)
 {
   auto start_time = std::chrono::steady_clock::now();
   if (m_verbose.load() ) {

--- a/src/server/request_context.cpp
+++ b/src/server/request_context.cpp
@@ -92,16 +92,16 @@ RequestContext::RequestContext(struct MHD_Connection* connection,
 RequestContext::~RequestContext()
 {}
 
-int RequestContext::fill_header(void *__this, enum MHD_ValueKind kind,
-                                 const char *key, const char *value)
+MHD_Result RequestContext::fill_header(void *__this, enum MHD_ValueKind kind,
+                                       const char *key, const char *value)
 {
   RequestContext *_this = static_cast<RequestContext*>(__this);
   _this->headers[key] = value;
   return MHD_YES;
 }
 
-int RequestContext::fill_argument(void *__this, enum MHD_ValueKind kind,
-                                   const char *key, const char* value)
+MHD_Result RequestContext::fill_argument(void *__this, enum MHD_ValueKind kind,
+                                         const char *key, const char* value)
 {
   RequestContext *_this = static_cast<RequestContext*>(__this);
   _this->arguments[key] = value == nullptr ? "" : value;

--- a/src/server/request_context.h
+++ b/src/server/request_context.h
@@ -30,7 +30,7 @@
 #include "byte_range.h"
 
 extern "C" {
-#include <microhttpd.h>
+#include "microhttpd_wrapper.h"
 }
 
 namespace kiwix {
@@ -98,8 +98,8 @@ class RequestContext {
     std::map<std::string, std::string> arguments;
 
   private: // functions
-    static int fill_header(void *, enum MHD_ValueKind, const char*, const char*);
-    static int fill_argument(void *, enum MHD_ValueKind, const char*, const char*);
+    static MHD_Result fill_header(void *, enum MHD_ValueKind, const char*, const char*);
+    static MHD_Result fill_argument(void *, enum MHD_ValueKind, const char*, const char*);
 };
 
 template<> std::string RequestContext::get_argument(const std::string& name) const;

--- a/src/server/response.cpp
+++ b/src/server/response.cpp
@@ -57,8 +57,8 @@ Response::Response(const std::string& root, bool verbose, bool withTaskbar, bool
 }
 
 
-static int print_key_value (void *cls, enum MHD_ValueKind kind,
-                            const char *key, const char *value)
+static MHD_Result print_key_value (void *cls, enum MHD_ValueKind kind,
+                                   const char *key, const char *value)
 {
   printf (" - %s: '%s'\n", key, value);
   return MHD_YES;
@@ -288,7 +288,7 @@ Response::create_mhd_response(const RequestContext& request)
   return nullptr;
 }
 
-int Response::send(const RequestContext& request, MHD_Connection* connection)
+MHD_Result Response::send(const RequestContext& request, MHD_Connection* connection)
 {
   MHD_Response* response = create_mhd_response(request);
 

--- a/src/server/response.h
+++ b/src/server/response.h
@@ -29,7 +29,7 @@
 #include "etag.h"
 
 extern "C" {
-#include <microhttpd.h>
+#include "microhttpd_wrapper.h"
 }
 
 namespace kiwix {
@@ -48,7 +48,7 @@ class Response {
     Response(const std::string& root, bool verbose, bool withTaskbar, bool withLibraryButton, bool blockExternalLinks);
     ~Response() = default;
 
-    int send(const RequestContext& request, MHD_Connection* connection);
+    MHD_Result send(const RequestContext& request, MHD_Connection* connection);
 
     void set_template(const std::string& template_str, kainjow::mustache::data data);
     void set_content(const std::string& content);

